src/encoder.rs

```diff
@@ -90,80 +78,63 @@ impl AsyncRead for Encoder {
-}
+#[derive(Debug)]
+pub struct Sender(async_channel::Sender<Vec<u8>>);
 
 /// Create a new SSE encoder.
 pub fn encode() -> (Sender, Encoder) {
-    let (sender, receiver) = sync::channel(1);
-    let disconnected = Arc::new(AtomicBool::new(false));
-
+    let (sender, receiver) = async_channel::bounded(1);
     let encoder = Encoder {
         receiver,
         buf: None,
         cursor: 0,
-        disconnected: disconnected.clone(),
     };
-
-    let sender = Sender {
-        sender,
-        disconnected,
-    };
-
-    (sender, encoder)
+    (Sender(sender), encoder)
 }
 
-/// An error that represents that the [Encoder] has been dropped.
-#[derive(Debug, Eq, PartialEq)]
-pub struct DisconnectedError;
-impl std::error::Error for DisconnectedError {}
-impl std::fmt::Display for DisconnectedError {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "Disconnected")
+impl Sender {
+    async fn inner_send(&self, bytes: impl Into<Vec<u8>>) -> io::Result<()> {
+        self.0
+            .send(bytes.into())
+            .await
+            .map_err(|_| io::Error::new(io::ErrorKind::ConnectionAborted, "sse disconnected"))
     }
-}
 
-#[must_use]
-impl Sender {
     /// Send a new message over SSE.
-    pub async fn send(
-        &self,
-        name: &str,
-        data: &str,
-        id: Option<&str>,
-    ) -> Result<(), DisconnectedError> {
-        if self.disconnected.load(Ordering::Relaxed) {
-            return Err(DisconnectedError);
-        }
-
+    pub async fn send(&self, name: &str, data: &str, id: Option<&str>) -> io::Result<()> {
         // Write the event name
         let msg = format!("event:{}\n", name);
-        self.sender.send(msg.into_bytes()).await;
+        self.inner_send(msg).await?;
 
         // Write the id
         if let Some(id) = id {
-            self.sender.send(format!("id:{}\n", id).into_bytes()).await;
+            self.inner_send(format!("id:{}\n", id)).await?;
         }
 
         // Write the data section, and end.
         let msg = format!("data:{}\n\n", data);
-        self.sender.send(msg.into_bytes()).await;
+        self.inner_send(msg).await?;
+
         Ok(())
     }
 
     /// Send a new "retry" message over SSE.
-    pub async fn send_retry(&self, dur: Duration, id: Option<&str>) {
+    pub async fn send_retry(&self, dur: Duration, id: Option<&str>) -> io::Result<()> {
         // Write the id
         if let Some(id) = id {
-            self.sender.send(format!("id:{}\n", id).into_bytes()).await;
+            self.inner_send(format!("id:{}\n", id)).await?;
         }
 
         // Write the retry section, and end.
         let dur = dur.as_secs_f64() as u64;
         let msg = format!("retry:{}\n\n", dur);
-        self.sender.send(msg.into_bytes()).await;
+        self.inner_send(msg).await?;
+        Ok(())
+    }
+}
+
+impl Clone for Sender {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
```

Comment:
Nice!

Reply:
This bit didn't actually change, although probably it should have. In the intermediate commit I changed this to `#[derive(Clone)]` because it's identical, but that got reverted

Reply:
switched back in f484fd446cb62f9b643e18ef49ad9b5403fc9872, lmk if there's a reason _not_ to use the derived clone impl

Reply:
@jbr :100: